From 9afee9cec35dd43be9142871b4a7c62b07424719 Mon Sep 17 00:00:00 2001
From: Marek Havrila <ma.havrila@gmail.com>
Date: Tue, 17 Feb 2026 17:24:02 +0100
Subject: [PATCH 09/12] Add loop to execute combs of weigths and alphas

Add weigths representing SQN and DDQN
Add variants of alpha
Run all combination of weights and alphas
Shorten and polish certain lines

diff --git a/src/CaserSMORL.py b/src/CaserSMORL.py
index 2ab3a7f..8d755ba 100644
--- a/src/CaserSMORL.py
+++ b/src/CaserSMORL.py
@@ -197,7 +197,7 @@ class Args:
     torch.backends.cudnn.deterministic = True
     dataset = 'rc15'
     eval_iter = 5000 if dataset == 'rc15' else 10000
-    epochs = 50
+    epochs = 3
     batch_size = 256
     hidden_factor = 64
     r_click = 0.2
@@ -216,8 +216,9 @@ class Args:
     results_to_file = True
     smorl_weights_set = [torch.Tensor([1.0, 1.0, 1.0]), torch.Tensor([0.0, 1.0, 0.0]),
                          torch.Tensor([0.0, 0.0, 1.0]), torch.Tensor([0.0, 1.0, 1.0]),
-                         torch.Tensor([1.0, 1.0, 0.0]), torch.Tensor([1.0, 0.0, 1.0])]
-    smorl_loss_mult = 1
+                         torch.Tensor([1.0, 1.0, 0.0]), torch.Tensor([1.0, 0.0, 1.0]),
+                         torch.Tensor([1.0, 0.0, 0.0]), torch.Tensor([0.0, 0.0, 0.0])]
+    smorl_loss_mults = [0.75, 1, 3, 10]
     discount = 0.5
     rel_disc_matrix = initialize_rel_disc_matrix(30)
 
@@ -228,170 +229,159 @@ if __name__ == '__main__':
 
     device = set_device(0)
     print('Using {} For Training'.format(torch.cuda.get_device_name()))
-    for smorl_weights in args.smorl_weights_set:
-        file_name = 'caser_smorl{}_acc{}_div{}_nov{}_weighted_q_vals.txt'.format(
-            str(args.smorl_loss_mult).replace('.', ''),
-            smorl_weights[0],
-            smorl_weights[1],
-            smorl_weights[2]
-        )
-        set_stdout(args.results_path, file_name, write_to_file=args.results_to_file)
-        sys.stdout.flush()
-        print("I'm starting caser_smorl{}_acc{}_div{}_nov{}_weighted_q_vals".format(
-            str(args.smorl_loss_mult).replace('.', ''),
-            smorl_weights[0],
-            smorl_weights[1],
-            smorl_weights[2]
-        ))
-        nov_rewards_dict = get_novelty_reward_dict(args.data_path)
-
-        state_size, item_num = get_stats(args.data_path)
-        train_loader = prepare_dataloader(args.data_path, args.batch_size)
-
-        caser1 = CaserSMORL(
-            hidden_size=args.hidden_factor,
-            item_num=item_num,
-            state_size=state_size,
-            num_filters=args.num_filters,
-            filter_sizes=args.filter_sizes,
-            dropout_rate=args.dropout_rate,
-            discount=args.discount,
-            div_embedding_matrix=args.div_emb_matrix,
-            smorl_loss_mult=args.smorl_loss_mult,
-            smorl_weights=smorl_weights.to(device),
-            novelty_rewards_dict=nov_rewards_dict,
-            device=device
-        )
-
-        caser2 = CaserSMORL(
-            hidden_size=args.hidden_factor,
-            item_num=item_num,
-            state_size=state_size,
-            num_filters=args.num_filters,
-            filter_sizes=args.filter_sizes,
-            dropout_rate=args.dropout_rate,
-            discount=args.discount,
-            div_embedding_matrix=args.div_emb_matrix,
-            smorl_loss_mult=args.smorl_loss_mult,
-            smorl_weights=smorl_weights.to(device),
-            novelty_rewards_dict=nov_rewards_dict,
-            device=device
-        )
 
-        print("Model:")
-        print(caser1)
-        print("TORCHINFO:")
-        summary(caser1)
-        print("Model ':")
-        print(caser2)
-        print("TORCHINFO:")
-        summary(caser2)
-
-        criterion = nn.CrossEntropyLoss()
-        params1 = list(caser1.parameters())
-        optimizer1 = torch.optim.Adam(params1, lr=args.lr)
-
-        params2 = list(caser2.parameters())
-        optimizer2 = torch.optim.Adam(params2, lr=args.lr)
-
-        caser1.to(device)
-        caser2.to(device)
-
-        reward_click = torch.repeat_interleave(torch.Tensor([args.r_click]), args.batch_size)
-        reward_buy = torch.repeat_interleave(torch.Tensor([args.r_buy]), args.batch_size)
-
-        # Start training loop
-        epoch_times = []
-        total_step = 0
-        for epoch in range(0, args.epochs):
-            caser1.train()
-            caser2.train()
-            start_time = time.time()
-            for state, len_state, action, next_state, len_next_state, is_buy, is_done in train_loader:
-                pointer = np.random.randint(0, 2)
-                if pointer == 0:
-                    main_QN = caser1
-                    target_QN = caser2
-                    optimizer = optimizer1
-                    model_name = 'Caser SMORL 1'
-                else:
-                    main_QN = caser2
-                    target_QN = caser1
-                    optimizer = optimizer2
-                    model_name = 'Caser SMORL 2'
-
-                main_QN.zero_grad()
-                target_QN.zero_grad()
-
-                target_Qs = target_QN.forward_next_states(
-                    next_state.to(device),
-                    len_next_state.to(device)
-                )
-                target_Qs_selector = main_QN.forward_next_states(
-                    next_state.to(device),
-                    len_next_state.to(device)
-                )
-                acc_reward = torch.where(is_buy > 0, reward_buy, reward_click)
-                supervised_out, smorl_loss = main_QN(
-                    state.to(device).long(),
-                    len_state.to(device).long(),
-                    action.to(device),
-                    acc_reward.to(device),
-                    target_Qs.to(device),
-                    target_Qs_selector.to(device),
-                    is_done.to(device)
-                )
-                supervised_loss = criterion(supervised_out, action.to(device).long())
-                loss = supervised_loss + smorl_loss
-                scalar_loss = loss.item()
-
-                optimizer.zero_grad()
-                loss.backward()
-                optimizer.step()
-
-                total_step += 1
-
-                if total_step % 200 == 0:
-                    print('Model is ', model_name)
-                    print('Supervised loss is %.3f, SMORL loss is %.3f' % (supervised_loss.item(), smorl_loss))
-                    print('Epoch {}......Step: {}....... Loss: {}'.format(epoch, total_step, scalar_loss))
-                    sys.stdout.flush()
-
-                if total_step % args.eval_iter == 0:
-                    print('Evaluating Main Model')
-                    val_acc_main, predictions_main = evaluate(main_QN, args, 'val', state_size, item_num, device,
-                                                              args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    val_acc_main, predictions_main = evaluate(main_QN, args, 'test', state_size, item_num, device,
-                                                              args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    predictions_main.to_csv(
-                        '{}pred_main_acc{}_div{}_nov{}_epoch{}_step{}.tsv'.format(args.results_path, smorl_weights[0],
-                                                                                  smorl_weights[1], smorl_weights[2],
-                                                                                  epoch, total_step), sep='\t',
-                        index=False)
-                    print("I've saved pred main at {}pred_main_acc{}_div{}_nov{}_epoch{}_step{}.tsv".format(
-                        args.results_path, smorl_weights[0], smorl_weights[1], smorl_weights[2], epoch, total_step))
-                    main_QN.train()
-                    print('Current accuracy of main model: ', val_acc_main)
-                    print('Evaluating Target Model')
-                    val_acc_target, predictions_target = evaluate(target_QN, args, 'val', state_size, item_num, device,
+    for smorl_loss_mult in args.smorl_loss_mults:
+        for smorl_weights in args.smorl_weights_set:
+            file_name = 'caser_smorl_{}_acc{}_div{}_nov{}.txt'.format(
+                str(smorl_loss_mult).replace('.', ''),
+                smorl_weights[0],
+                smorl_weights[1],
+                smorl_weights[2]
+            )
+            experiment_name = file_name[:-4]
+            set_stdout(args.results_path, file_name, write_to_file=args.results_to_file)
+            sys.stdout.flush()
+            print(f"I'm starting {experiment_name}")
+            nov_rewards_dict = get_novelty_reward_dict(args.data_path)
+
+            state_size, item_num = get_stats(args.data_path)
+            train_loader = prepare_dataloader(args.data_path, args.batch_size)
+
+            caser1 = CaserSMORL(
+                hidden_size=args.hidden_factor,
+                item_num=item_num,
+                state_size=state_size,
+                num_filters=args.num_filters,
+                filter_sizes=args.filter_sizes,
+                dropout_rate=args.dropout_rate,
+                discount=args.discount,
+                div_embedding_matrix=args.div_emb_matrix,
+                smorl_loss_mult=smorl_loss_mult,
+                smorl_weights=smorl_weights.to(device),
+                novelty_rewards_dict=nov_rewards_dict,
+                device=device
+            )
+
+            caser2 = CaserSMORL(
+                hidden_size=args.hidden_factor,
+                item_num=item_num,
+                state_size=state_size,
+                num_filters=args.num_filters,
+                filter_sizes=args.filter_sizes,
+                dropout_rate=args.dropout_rate,
+                discount=args.discount,
+                div_embedding_matrix=args.div_emb_matrix,
+                smorl_loss_mult=smorl_loss_mult,
+                smorl_weights=smorl_weights.to(device),
+                novelty_rewards_dict=nov_rewards_dict,
+                device=device
+            )
+
+            print("Model:")
+            print(caser1)
+            print("TORCHINFO:")
+            summary(caser1)
+            print("Model ':")
+            print(caser2)
+            print("TORCHINFO:")
+            summary(caser2)
+
+            criterion = nn.CrossEntropyLoss()
+            params1 = list(caser1.parameters())
+            optimizer1 = torch.optim.Adam(params1, lr=args.lr)
+
+            params2 = list(caser2.parameters())
+            optimizer2 = torch.optim.Adam(params2, lr=args.lr)
+
+            caser1.to(device)
+            caser2.to(device)
+
+            reward_click = torch.repeat_interleave(torch.Tensor([args.r_click]), args.batch_size)
+            reward_buy = torch.repeat_interleave(torch.Tensor([args.r_buy]), args.batch_size)
+
+            # Start training loop
+            epoch_times = []
+            total_step = 0
+            for epoch in range(0, args.epochs):
+                caser1.train()
+                caser2.train()
+                start_time = time.time()
+                for state, len_state, action, next_state, len_next_state, is_buy, is_done in train_loader:
+                    pointer = np.random.randint(0, 2)
+                    if pointer == 0:
+                        main_QN = caser1
+                        target_QN = caser2
+                        optimizer = optimizer1
+                        model_name = 'Caser SMORL 1'
+                    else:
+                        main_QN = caser2
+                        target_QN = caser1
+                        optimizer = optimizer2
+                        model_name = 'Caser SMORL 2'
+
+                    main_QN.zero_grad()
+                    target_QN.zero_grad()
+
+                    target_Qs = target_QN.forward_next_states(
+                        next_state.to(device),
+                        len_next_state.to(device)
+                    )
+                    target_Qs_selector = main_QN.forward_next_states(
+                        next_state.to(device),
+                        len_next_state.to(device)
+                    )
+                    acc_reward = torch.where(is_buy > 0, reward_buy, reward_click)
+                    supervised_out, smorl_loss = main_QN(
+                        state.to(device).long(),
+                        len_state.to(device).long(),
+                        action.to(device),
+                        acc_reward.to(device),
+                        target_Qs.to(device),
+                        target_Qs_selector.to(device),
+                        is_done.to(device)
+                    )
+                    supervised_loss = criterion(supervised_out, action.to(device).long())
+                    loss = supervised_loss + smorl_loss
+                    scalar_loss = loss.item()
+
+                    optimizer.zero_grad()
+                    loss.backward()
+                    optimizer.step()
+
+                    total_step += 1
+
+                    if total_step % 200 == 0:
+                        print('Model is ', model_name)
+                        print('Supervised loss is %.3f, SMORL loss is %.3f' % (supervised_loss.item(), smorl_loss))
+                        print('Epoch {}......Step: {}....... Loss: {}'.format(epoch, total_step, scalar_loss))
+                        sys.stdout.flush()
+
+                    if total_step % args.eval_iter == 0:
+                        print('Evaluating Main Model')
+                        val_acc_main, predictions_main = evaluate(main_QN, args, 'val', state_size, item_num, device,
                                                                   args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    val_acc_target, predictions_target = evaluate(target_QN, args, 'test', state_size, item_num, device,
+                        val_acc_main, predictions_main = evaluate(main_QN, args, 'test', state_size, item_num, device,
                                                                   args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    predictions_target.to_csv(
-                        '{}pred_target_acc{}_div{}_nov{}_epoch{}_step{}.tsv'.format(args.results_path, smorl_weights[0],
-                                                                                    smorl_weights[1], smorl_weights[2],
-                                                                                    epoch, total_step), sep='\t',
-                        index=False)
-                    print("I've saved pred target at {}pred_target_acc{}_div{}_nov{}_epoch{}_step{}.tsv".format(
-                        args.results_path,
-                        smorl_weights[0], smorl_weights[1], smorl_weights[2], epoch, total_step))
-                    target_QN.train()
-                    print('Current accuracy of target model: ', val_acc_target)
-
-            current_time = time.time()
-            print('Epoch {}/{} Done'.format(epoch, args.epochs))
-            print('Total Time Elapsed: {} seconds'.format(str(current_time - start_time)))
-            epoch_times.append(current_time - start_time)
-        print('Total Training Time: {} seconds'.format(str(sum(epoch_times))))
-        sys.stdout.close()
-        sys.stdout = sys.__stdout__
+                        predictions_main.to_csv(f'{args.results_path}pred_main_{experiment_name}_epoch{epoch}_step{total_step}.tsv',
+                                                sep='\t', index=False)
+                        print(f"I've saved pred main at {args.results_path}pred_main_{experiment_name}_epoch{epoch}_step{total_step}.tsv") 
+                        main_QN.train()
+                        print('Current accuracy of main model: ', val_acc_main)
+                        print('Evaluating Target Model')
+                        val_acc_target, predictions_target = evaluate(target_QN, args, 'val', state_size, item_num, device,
+                                                                      args.div_emb_matrix, skip=args.skip, RL=args.rl)
+                        val_acc_target, predictions_target = evaluate(target_QN, args, 'test', state_size, item_num, device,
+                                                                      args.div_emb_matrix, skip=args.skip, RL=args.rl)
+                        predictions_target.to_csv(f'{args.results_path}pred_targetd_{experiment_name}_epoch{epoch}_step{total_step}.tsv',
+                                                  sep='\t', index=False)
+                        print(f"I've saved pred target at {args.results_path}pred_target_{experiment_name}_epoch{epoch}_step{total_step}.tsv")
+                        target_QN.train()
+                        print('Current accuracy of target model: ', val_acc_target)
+
+                current_time = time.time()
+                print('Epoch {}/{} Done'.format(epoch, args.epochs))
+                print('Total Time Elapsed: {} seconds'.format(str(current_time - start_time)))
+                epoch_times.append(current_time - start_time)
+            print('Total Training Time: {} seconds'.format(str(sum(epoch_times))))
+            sys.stdout.close()
+            sys.stdout = sys.__stdout__
diff --git a/src/GRUSMORL.py b/src/GRUSMORL.py
index 46934c6..762e4bd 100644
--- a/src/GRUSMORL.py
+++ b/src/GRUSMORL.py
@@ -155,8 +155,9 @@ class Args:
     results_to_file = True
     smorl_weights_set = [torch.Tensor([1.0, 1.0, 1.0]), torch.Tensor([0.0, 1.0, 0.0]),
                          torch.Tensor([0.0, 0.0, 1.0]), torch.Tensor([0.0, 1.0, 1.0]),
-                         torch.Tensor([1.0, 1.0, 0.0]), torch.Tensor([1.0, 0.0, 1.0])]
-    smorl_loss_mult = 1
+                         torch.Tensor([1.0, 1.0, 0.0]), torch.Tensor([1.0, 0.0, 1.0]),
+                         torch.Tensor([1.0, 0.0, 0.0]), torch.Tensor([0.0, 0.0, 0.0])]
+    smorl_loss_mults = [0.75, 1, 3, 10]
     discount = 0.5
     rel_disc_matrix = initialize_rel_disc_matrix(30)
 
@@ -167,152 +168,150 @@ if __name__ == '__main__':
 
     device = set_device(0)
     print('Using {} For Training'.format(torch.cuda.get_device_name()))
-    for smorl_weights in args.smorl_weights_set:
-        file_name = 'gru_smorl{}_acc{}_div{}_nov{}_weighted_q_vals.txt'.format(
-            str(args.smorl_loss_mult).replace('.', ''),
-            smorl_weights[0],
-            smorl_weights[1],
-            smorl_weights[2])
-        set_stdout(args.results_path, file_name, write_to_file=args.results_to_file)
-        sys.stdout.flush()
-        print("I'm starting gru_smorl{}_acc{}_div{}_nov{}_weighted_q_vals".format(
-            str(args.smorl_loss_mult).replace('.', ''),
-            smorl_weights[0],
-            smorl_weights[1],
-            smorl_weights[2]
-        ))
-
-        nov_rewards_dict = get_novelty_reward_dict(args.data_path)
-
-        state_size, item_num = get_stats(args.data_path)
-        train_loader = prepare_dataloader(args.data_path, args.batch_size)
-
-        gru1 = GRUSMORL(
-            hidden_size=args.hidden_factor,
-            item_num=item_num,
-            state_size=state_size,
-            div_embedding_matrix=args.div_emb_matrix,
-            discount=args.discount,
-            smorl_loss_mult=args.smorl_loss_mult,
-            smorl_weights=smorl_weights.to(device),
-            novelty_rewards_dict=nov_rewards_dict,
-            device=device
-        )
-
-        gru2 = GRUSMORL(
-            hidden_size=args.hidden_factor,
-            item_num=item_num,
-            state_size=state_size,
-            div_embedding_matrix=args.div_emb_matrix,
-            discount=args.discount,
-            smorl_loss_mult=args.smorl_loss_mult,
-            smorl_weights=smorl_weights.to(device),
-            novelty_rewards_dict=nov_rewards_dict,
-            device=device
-        )
-
-        print("Model:")
-        print(gru1)
-        print("TORCHINFO:")
-        summary(gru1)
-        print("Model ':")
-        print(gru2)
-        print("TORCHINFO:")
-        summary(gru2) 
-
-        criterion = nn.CrossEntropyLoss()
-        params1 = list(gru1.parameters())
-        optimizer1 = torch.optim.Adam([param for param in params1 if param.requires_grad], lr=args.lr)
-
-        params2 = list(gru2.parameters())
-        optimizer2 = torch.optim.Adam([param for param in params2 if param.requires_grad], lr=args.lr)
-
-        gru1.to(device)
-        gru2.to(device)
-
-        reward_click = torch.repeat_interleave(torch.Tensor([args.r_click]), args.batch_size)
-        reward_buy = torch.repeat_interleave(torch.Tensor([args.r_buy]), args.batch_size)
-
-        # Start training loop
-        epoch_times = []
-        total_step = 0
-        for epoch in range(0, args.epochs):
-            gru1.train()
-            gru2.train()
-            start_time = time.time()
-            for state, len_state, action, next_state, len_next_state, is_buy, is_done in train_loader:
-                pointer = np.random.randint(0, 2)
-                if pointer == 0:
-                    main_QN = gru1
-                    target_QN = gru2
-                    optimizer = optimizer1
-                    model_name = 'GRU SMORL 1'
-                else:
-                    main_QN = gru2
-                    target_QN = gru1
-                    optimizer = optimizer2
-                    model_name = 'GRU SMORL 2'
-
-                main_QN.zero_grad()
-                target_QN.zero_grad()
-
-                target_Qs = target_QN.forward_next_states(
-                    next_state.to(device),
-                    len_next_state
-                )
-                target_Qs_selector = main_QN.forward_next_states(
-                    next_state.to(device),
-                    len_next_state
-                )
-                acc_reward = torch.where(is_buy > 0, reward_buy, reward_click)
-                supervised_out, smorl_loss = main_QN(
-                    state.to(device).long(),
-                    len_state.long(),
-                    action.to(device),
-                    acc_reward.to(device),
-                    target_Qs.to(device),
-                    target_Qs_selector.to(device),
-                    is_done.to(device)
-                )
-                supervised_loss = criterion(supervised_out, action.to(device).long())
-                loss = supervised_loss + smorl_loss
-                scalar_loss = loss.item()
-
-                optimizer.zero_grad()
-                loss.backward()
-                optimizer.step()
-
-                total_step += 1
-
-                if total_step % 200 == 0:
-                    print('Model is ', model_name)
-                    print('Supervised loss is %.3f, SMORL loss is %.3f' % (supervised_loss.item(), smorl_loss))
-                    print('Epoch {}......Step: {}....... Loss: {}'.format(epoch, total_step, scalar_loss))
-                    sys.stdout.flush()
-
-                if total_step % args.eval_iter == 0:
-                    print('Evaluating Main Model')
-                    val_acc_main, predictions_main = evaluate(main_QN, args, 'val', state_size, item_num, device, args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    val_acc_main, predictions_main = evaluate(main_QN, args, 'test', state_size, item_num, device,
-                                                              args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    predictions_main.to_csv('{}pred_main_acc{}_div{}_nov{}_epoch{}_step{}.tsv'.format(args.results_path, smorl_weights[0],smorl_weights[1],smorl_weights[2], epoch, total_step), sep='\t', index=False)
-                    print("I've saved pred main at {}pred_main_acc{}_div{}_nov{}_epoch{}_step{}.tsv".format(args.results_path, smorl_weights[0],smorl_weights[1],smorl_weights[2], epoch, total_step))
-                    main_QN.train()
-                    print('Current accuracy of main model: ', val_acc_main)
-                    print('Evaluating Target Model')
-                    val_acc_target, predictions_target = evaluate(target_QN, args, 'val', state_size, item_num, device, args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    val_acc_target, predictions_target = evaluate(target_QN, args, 'test', state_size, item_num, device,
+    for smorl_loss_mult in args.smorl_loss_mults:
+        for smorl_weights in args.smorl_weights_set:
+            file_name = 'gru_smorl_{}_acc{}_div{}_nov{}'.format(
+                str(smorl_loss_mult).replace('.', ''),
+                smorl_weights[0],
+                smorl_weights[1],
+                smorl_weights[2])
+            experiment_name = file_name[:-4]
+            set_stdout(args.results_path, file_name, write_to_file=args.results_to_file)
+            sys.stdout.flush()
+            print(f"I'm starting {experiment_name}") 
+
+            nov_rewards_dict = get_novelty_reward_dict(args.data_path)
+
+            state_size, item_num = get_stats(args.data_path)
+            train_loader = prepare_dataloader(args.data_path, args.batch_size)
+
+            gru1 = GRUSMORL(
+                hidden_size=args.hidden_factor,
+                item_num=item_num,
+                state_size=state_size,
+                div_embedding_matrix=args.div_emb_matrix,
+                discount=args.discount,
+                smorl_loss_mult=smorl_loss_mult,
+                smorl_weights=smorl_weights.to(device),
+                novelty_rewards_dict=nov_rewards_dict,
+                device=device
+            )
+
+            gru2 = GRUSMORL(
+                hidden_size=args.hidden_factor,
+                item_num=item_num,
+                state_size=state_size,
+                div_embedding_matrix=args.div_emb_matrix,
+                discount=args.discount,
+                smorl_loss_mult=smorl_loss_mult,
+                smorl_weights=smorl_weights.to(device),
+                novelty_rewards_dict=nov_rewards_dict,
+                device=device
+            )
+
+            print("Model:")
+            print(gru1)
+            print("TORCHINFO:")
+            summary(gru1)
+            print("Model ':")
+            print(gru2)
+            print("TORCHINFO:")
+            summary(gru2) 
+
+            criterion = nn.CrossEntropyLoss()
+            params1 = list(gru1.parameters())
+            optimizer1 = torch.optim.Adam([param for param in params1 if param.requires_grad], lr=args.lr)
+
+            params2 = list(gru2.parameters())
+            optimizer2 = torch.optim.Adam([param for param in params2 if param.requires_grad], lr=args.lr)
+
+            gru1.to(device)
+            gru2.to(device)
+
+            reward_click = torch.repeat_interleave(torch.Tensor([args.r_click]), args.batch_size)
+            reward_buy = torch.repeat_interleave(torch.Tensor([args.r_buy]), args.batch_size)
+
+            # Start training loop
+            epoch_times = []
+            total_step = 0
+            for epoch in range(0, args.epochs):
+                gru1.train()
+                gru2.train()
+                start_time = time.time()
+                for state, len_state, action, next_state, len_next_state, is_buy, is_done in train_loader:
+                    pointer = np.random.randint(0, 2)
+                    if pointer == 0:
+                        main_QN = gru1
+                        target_QN = gru2
+                        optimizer = optimizer1
+                        model_name = 'GRU SMORL 1'
+                    else:
+                        main_QN = gru2
+                        target_QN = gru1
+                        optimizer = optimizer2
+                        model_name = 'GRU SMORL 2'
+
+                    main_QN.zero_grad()
+                    target_QN.zero_grad()
+
+                    target_Qs = target_QN.forward_next_states(
+                        next_state.to(device),
+                        len_next_state
+                    )
+                    target_Qs_selector = main_QN.forward_next_states(
+                        next_state.to(device),
+                        len_next_state
+                    )
+                    acc_reward = torch.where(is_buy > 0, reward_buy, reward_click)
+                    supervised_out, smorl_loss = main_QN(
+                        state.to(device).long(),
+                        len_state.long(),
+                        action.to(device),
+                        acc_reward.to(device),
+                        target_Qs.to(device),
+                        target_Qs_selector.to(device),
+                        is_done.to(device)
+                    )
+                    supervised_loss = criterion(supervised_out, action.to(device).long())
+                    loss = supervised_loss + smorl_loss
+                    scalar_loss = loss.item()
+
+                    optimizer.zero_grad()
+                    loss.backward()
+                    optimizer.step()
+
+                    total_step += 1
+
+                    if total_step % 200 == 0:
+                        print('Model is ', model_name)
+                        print('Supervised loss is %.3f, SMORL loss is %.3f' % (supervised_loss.item(), smorl_loss))
+                        print('Epoch {}......Step: {}....... Loss: {}'.format(epoch, total_step, scalar_loss))
+                        sys.stdout.flush()
+
+                    if total_step % args.eval_iter == 0:
+                        print('Evaluating Main Model')
+                        val_acc_main, predictions_main = evaluate(main_QN, args, 'val', state_size, item_num, device, args.div_emb_matrix, skip=args.skip, RL=args.rl)
+                        val_acc_main, predictions_main = evaluate(main_QN, args, 'test', state_size, item_num, device,
                                                                   args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    predictions_target.to_csv('{}pred_target_acc{}_div{}_nov{}_epoch{}_step{}.tsv'.format(args.results_path,smorl_weights[0], smorl_weights[1],smorl_weights[2], epoch, total_step), sep='\t',index=False)
-                    print("I've saved pred target at {}pred_target_acc{}_div{}_nov{}_epoch{}_step{}.tsv".format(args.results_path,
-                        smorl_weights[0], smorl_weights[1], smorl_weights[2], epoch, total_step))
-                    target_QN.train()
-                    print('Current accuracy of target model: ', val_acc_target)
-
-            current_time = time.time()
-            print('Epoch {}/{} Done'.format(epoch, args.epochs))
-            print('Total Time Elapsed: {} seconds'.format(str(current_time - start_time)))
-            epoch_times.append(current_time - start_time)
-        print('Total Training Time: {} seconds'.format(str(sum(epoch_times))))
-        sys.stdout.close()
-        sys.stdout = sys.__stdout__
+                        predictions_main.to_csv(f'{args.results_path}pred_main_{experiment_name}_epoch{epoch}_step{total_step}.tsv',
+                                                sep='\t', index=False)
+                        print(f"I've saved pred main at {args.results_path}pred_main_{experiment_name}_epoch{epoch}_step{total_step}.tsv")
+                        main_QN.train()
+                        print('Current accuracy of main model: ', val_acc_main)
+                        print('Evaluating Target Model')
+                        val_acc_target, predictions_target = evaluate(target_QN, args, 'val', state_size, item_num, device, args.div_emb_matrix, skip=args.skip, RL=args.rl)
+                        val_acc_target, predictions_target = evaluate(target_QN, args, 'test', state_size, item_num, device,
+                                                                      args.div_emb_matrix, skip=args.skip, RL=args.rl)
+                        predictions_target.to_csv(f'{args.results_path}pred_targetd_{experiment_name}_epoch{epoch}_step{total_step}.tsv',
+                                                  sep='\t', index=False)
+                        print(f"I've saved pred target at {args.results_path}pred_target_{experiment_name}_epoch{epoch}_step{total_step}.tsv")
+                        target_QN.train()
+                        print('Current accuracy of target model: ', val_acc_target)
+
+                current_time = time.time()
+                print('Epoch {}/{} Done'.format(epoch, args.epochs))
+                print('Total Time Elapsed: {} seconds'.format(str(current_time - start_time)))
+                epoch_times.append(current_time - start_time)
+            print('Total Training Time: {} seconds'.format(str(sum(epoch_times))))
+            sys.stdout.close()
+            sys.stdout = sys.__stdout__
diff --git a/src/NextItNetSMORL.py b/src/NextItNetSMORL.py
index a1c4f11..e136131 100644
--- a/src/NextItNetSMORL.py
+++ b/src/NextItNetSMORL.py
@@ -169,7 +169,8 @@ class Args:
     results_to_file = True
     smorl_weights_set = [torch.Tensor([1.0, 1.0, 1.0]), torch.Tensor([0.0, 1.0, 0.0]),
                          torch.Tensor([0.0, 0.0, 1.0]), torch.Tensor([0.0, 1.0, 1.0]),
-                         torch.Tensor([1.0, 1.0, 0.0]), torch.Tensor([1.0, 0.0, 1.0])]
+                         torch.Tensor([1.0, 1.0, 0.0]), torch.Tensor([1.0, 0.0, 1.0]),
+                         torch.Tensor([1.0, 0.0, 0.0]), torch.Tensor([0.0, 0.0, 0.0])]
     smorl_loss_mult = 1
     discount = 0.5
     rel_disc_matrix = initialize_rel_disc_matrix(30)
@@ -188,14 +189,10 @@ if __name__ == '__main__':
             smorl_weights[1],
             smorl_weights[2]
         )
+        experiment_name = file_name[:-4]
         set_stdout(args.results_path, file_name, write_to_file=args.results_to_file)
         sys.stdout.flush()
-        print("I'm starting nextitnet_smorl{}_acc{}_div{}_nov{}_weighted_q_vals".format(
-            str(args.smorl_loss_mult).replace('.', ''),
-            smorl_weights[0],
-            smorl_weights[1],
-            smorl_weights[2]
-        ))
+        print(f"I'm starting {experiment_name}")
         nov_rewards_dict = get_novelty_reward_dict(args.data_path)
 
 
@@ -317,27 +314,21 @@ if __name__ == '__main__':
                                                               args.div_emb_matrix, skip=args.skip, RL=args.rl)
                     val_acc_main, predictions_main = evaluate(main_QN, args, 'test', state_size, item_num, device,
                                                               args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    predictions_main.to_csv(
-                        '{}pred_main_acc{}_div{}_nov{}_epoch{}_step{}.tsv'.format(args.results_path, smorl_weights[0],
-                                                                                  smorl_weights[1], smorl_weights[2],
-                                                                                  epoch, total_step), sep='\t',
-                        index=False)
-                    print("I've saved pred main at {}pred_main_acc{}_div{}_nov{}_epoch{}_step{}.tsv".format(
-                        args.results_path, smorl_weights[0], smorl_weights[1], smorl_weights[2], epoch, total_step))
+
+
+                    predictions_main.to_csv(f'{args.results_path}pred_main_{experiment_name}_epoch{epoch}_step{total_step}.tsv',
+                                                sep='\t', index=False)
+                    print(f"I've saved pred main at {args.results_path}pred_main_{experiment_name}_epoch{epoch}_step{total_step}.tsv") 
+                    main_QN.train()
                     print('Current accuracy of main model: ', val_acc_main)
                     print('Evaluating Target Model')
                     val_acc_target, predictions_target = evaluate(target_QN, args, 'val', state_size, item_num, device,
                                                                   args.div_emb_matrix, skip=args.skip, RL=args.rl)
                     val_acc_target, predictions_target = evaluate(target_QN, args, 'test', state_size, item_num, device,
                                                                   args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    predictions_target.to_csv(
-                        '{}pred_target_acc{}_div{}_nov{}_epoch{}_step{}.tsv'.format(args.results_path, smorl_weights[0],
-                                                                                    smorl_weights[1], smorl_weights[2],
-                                                                                    epoch, total_step), sep='\t',
-                        index=False)
-                    print("I've saved pred target at {}pred_target_acc{}_div{}_nov{}_epoch{}_step{}.tsv".format(
-                        args.results_path,
-                        smorl_weights[0], smorl_weights[1], smorl_weights[2], epoch, total_step))
+                    predictions_target.to_csv(f'{args.results_path}pred_targetd_{experiment_name}_epoch{epoch}_step{total_step}.tsv',
+                                                  sep='\t', index=False)
+                    print(f"I've saved pred target at {args.results_path}pred_target_{experiment_name}_epoch{epoch}_step{total_step}.tsv")
                     target_QN.train()
                     print('Current accuracy of target model: ', val_acc_target)
             current_time = time.time()
diff --git a/src/SASRecSMORL.py b/src/SASRecSMORL.py
index 381a5b8..1e5d7e4 100644
--- a/src/SASRecSMORL.py
+++ b/src/SASRecSMORL.py
@@ -180,8 +180,9 @@ class Args:
     discount = 0.5
     smorl_weights_set = [torch.Tensor([1.0, 1.0, 1.0]), torch.Tensor([0.0, 1.0, 0.0]),
                          torch.Tensor([0.0, 0.0, 1.0]), torch.Tensor([0.0, 1.0, 1.0]),
-                         torch.Tensor([1.0, 1.0, 0.0]), torch.Tensor([1.0, 0.0, 1.0])]
-    smorl_loss_mult = 1
+                         torch.Tensor([1.0, 1.0, 0.0]), torch.Tensor([1.0, 0.0, 1.0]),
+                         torch.Tensor([1.0, 0.0, 0.0]), torch.Tensor([0.0, 0.0, 0.0])]
+    smorl_loss_mults = [1, 3, 10]
     rel_disc_matrix = initialize_rel_disc_matrix(30)
 
 
@@ -191,162 +192,151 @@ if __name__ == '__main__':
 
     device = set_device(2)
     print('Using {} For Training'.format(torch.cuda.get_device_name(0)))
-    for smorl_weights in args.smorl_weights_set:
-        file_name = 'sasrec_smorl{}_acc{}_div{}_nov{}_weighted_q_vals.txt'.format(
-            str(args.smorl_loss_mult).replace('.', ''),
-            smorl_weights[0],
-            smorl_weights[1],
-            smorl_weights[2])
-        set_stdout(args.results_path, file_name, write_to_file=args.results_to_file)
-        sys.stdout.flush()
-        print("I'm starting sasrec_smorl{}_acc{}_div{}_nov{}_weighted_q_vals".format(
-            str(args.smorl_loss_mult).replace('.', ''),
-            smorl_weights[0],
-            smorl_weights[1],
-            smorl_weights[2]
-        ))
-        nov_rewards_dict = get_novelty_reward_dict(args.data_path)
-
-        state_size, item_num = get_stats(args.data_path)
-        train_loader = prepare_dataloader(args.data_path, args.batch_size)
-        sasrec1 = SASRecSMORL(
-            hidden_size=args.hidden_factor,
-            item_num=item_num,
-            state_size=state_size,
-            dropout=args.dropout_rate,
-            div_embedding_matrix= args.div_emb_matrix,
-            discount=args.discount,
-            smorl_loss_mult=args.smorl_loss_mult,
-            smorl_weights=smorl_weights.to(device),
-            novelty_rewards_dict=nov_rewards_dict,
-            device=device
-        )
-
-        sasrec2 = SASRecSMORL(
-            hidden_size=args.hidden_factor,
-            item_num=item_num,
-            state_size=state_size,
-            dropout=args.dropout_rate,
-            div_embedding_matrix=args.div_emb_matrix,
-            discount=args.discount,
-            smorl_loss_mult=args.smorl_loss_mult,
-            smorl_weights=smorl_weights.to(device),
-            novelty_rewards_dict=nov_rewards_dict,
-            device=device
-        )
-
-        print("Model:")
-        print(sasrec1)
-        print("TORCHINFO:")
-        summary(sasrec1)
-        print("Model ':")
-        print(sasrec2)
-        print("TORCHINFO:")
-        summary(sasrec2)
-
-        criterion = nn.CrossEntropyLoss()
-        params1 = list(sasrec1.parameters())
-        optimizer1 = torch.optim.Adam(params1, lr=args.lr)
-
-        params2 = list(sasrec2.parameters())
-        optimizer2 = torch.optim.Adam(params2, lr=args.lr)
-
-        sasrec1.to(device)
-        sasrec2.to(device)
-
-        reward_click = torch.repeat_interleave(torch.Tensor([args.r_click]), args.batch_size)
-        reward_buy = torch.repeat_interleave(torch.Tensor([args.r_buy]), args.batch_size)
-
-        # Start training loop
-        epoch_times = []
-        total_step = 0
-        for epoch in range(0, args.epochs):
-            sasrec1.train()
-            sasrec2.train()
-            start_time = time.time()
-            for state, len_state, action, next_state, len_next_state, is_buy, is_done in train_loader:
-                pointer = np.random.randint(0, 2)
-                if pointer == 0:
-                    main_QN = sasrec1
-                    target_QN = sasrec2
-                    optimizer = optimizer1
-                    model_name = 'SASRec SMORL 1'
-                else:
-                    main_QN = sasrec2
-                    target_QN = sasrec1
-                    optimizer = optimizer2
-                    model_name = 'SASRec SMORL 2'
-
-                main_QN.zero_grad()
-                target_QN.zero_grad()
-
-                target_Qs = target_QN.forward_next_states(
-                    next_state.to(device),
-                    len_next_state.to(device)
-                )
-                target_Qs_selector = main_QN.forward_next_states(
-                    next_state.to(device),
-                    len_next_state.to(device)
-                )
-                acc_reward = torch.where(is_buy > 0, reward_buy, reward_click)
-                supervised_out, smorl_loss = main_QN(
-                    state.to(device).long(),
-                    len_state.to(device).long(),
-                    action.to(device),
-                    acc_reward.to(device),
-                    target_Qs.to(device),
-                    target_Qs_selector.to(device),
-                    is_done.to(device)
-                )
-                supervised_loss = criterion(supervised_out, action.to(device).long())
-                loss = supervised_loss + smorl_loss
-                scalar_loss = loss.item()
-
-                optimizer.zero_grad()
-                loss.backward()
-                optimizer.step()
-
-                total_step += 1
-
-                if total_step % 200 == 0:
-                    print('Model is ', model_name)
-                    print('Supervised loss is %.3f, SMORL loss is %.3f' % (supervised_loss.item(), smorl_loss))
-                    print('Epoch {}......Step: {}....... Loss: {}'.format(epoch, total_step, scalar_loss = loss.item()))
-                    sys.stdout.flush()
-
-                if total_step % args.eval_iter == 0:
-                    print('Evaluating Main Model')
-                    val_acc_main, predictions_main = evaluate(main_QN, args, 'val', state_size, item_num, device, args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    val_acc_main, predictions_main = evaluate(main_QN, args, 'test', state_size, item_num, device,
-                                                              args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    predictions_main.to_csv(
-                        '{}pred_main_acc{}_div{}_nov{}_epoch{}_step{}.tsv'.format(args.results_path, smorl_weights[0],
-                                                                                  smorl_weights[1], smorl_weights[2],
-                                                                                  epoch, total_step), sep='\t',
-                        index=False)
-                    print("I've saved pred main at {}pred_main_acc{}_div{}_nov{}_epoch{}_step{}.tsv".format(
-                        args.results_path, smorl_weights[0], smorl_weights[1], smorl_weights[2], epoch, total_step))
-                    main_QN.train()
-                    print('Current accuracy of main model: ', val_acc_main)
-                    print('Evaluating Target Model')
-                    val_acc_target, predictions_target = evaluate(target_QN, args, 'val', state_size, item_num, device, args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    val_acc_target, predictions_target = evaluate(target_QN, args, 'test', state_size, item_num, device,
+    for smorl_loss_mult in args.smorl_loss_mults:
+        for smorl_weights in args.smorl_weights_set:
+            file_name = 'sasrec_smorl_{}_acc{}_div{}_nov{}.txt'.format(
+                str(smorl_loss_mult).replace('.', ''),
+                smorl_weights[0],
+                smorl_weights[1],
+                smorl_weights[2])
+            experiment_name = file_name[:-4]
+            set_stdout(args.results_path, file_name, write_to_file=args.results_to_file)
+            sys.stdout.flush()
+            print(f"I'm starting {experiment_name}")
+            nov_rewards_dict = get_novelty_reward_dict(args.data_path)
+
+            state_size, item_num = get_stats(args.data_path)
+            train_loader = prepare_dataloader(args.data_path, args.batch_size)
+            sasrec1 = SASRecSMORL(
+                hidden_size=args.hidden_factor,
+                item_num=item_num,
+                state_size=state_size,
+                dropout=args.dropout_rate,
+                div_embedding_matrix= args.div_emb_matrix,
+                discount=args.discount,
+                smorl_loss_mult=smorl_loss_mult,
+                smorl_weights=smorl_weights.to(device),
+                novelty_rewards_dict=nov_rewards_dict,
+                device=device
+            )
+
+            sasrec2 = SASRecSMORL(
+                hidden_size=args.hidden_factor,
+                item_num=item_num,
+                state_size=state_size,
+                dropout=args.dropout_rate,
+                div_embedding_matrix=args.div_emb_matrix,
+                discount=args.discount,
+                smorl_loss_mult=smorl_loss_mult,
+                smorl_weights=smorl_weights.to(device),
+                novelty_rewards_dict=nov_rewards_dict,
+                device=device
+            )
+
+            print("Model:")
+            print(sasrec1)
+            print("TORCHINFO:")
+            summary(sasrec1)
+            print("Model ':")
+            print(sasrec2)
+            print("TORCHINFO:")
+            summary(sasrec2)
+
+            criterion = nn.CrossEntropyLoss()
+            params1 = list(sasrec1.parameters())
+            optimizer1 = torch.optim.Adam(params1, lr=args.lr)
+
+            params2 = list(sasrec2.parameters())
+            optimizer2 = torch.optim.Adam(params2, lr=args.lr)
+
+            sasrec1.to(device)
+            sasrec2.to(device)
+
+            reward_click = torch.repeat_interleave(torch.Tensor([args.r_click]), args.batch_size)
+            reward_buy = torch.repeat_interleave(torch.Tensor([args.r_buy]), args.batch_size)
+
+            # Start training loop
+            epoch_times = []
+            total_step = 0
+            for epoch in range(0, args.epochs):
+                sasrec1.train()
+                sasrec2.train()
+                start_time = time.time()
+                for state, len_state, action, next_state, len_next_state, is_buy, is_done in train_loader:
+                    pointer = np.random.randint(0, 2)
+                    if pointer == 0:
+                        main_QN = sasrec1
+                        target_QN = sasrec2
+                        optimizer = optimizer1
+                        model_name = 'SASRec SMORL 1'
+                    else:
+                        main_QN = sasrec2
+                        target_QN = sasrec1
+                        optimizer = optimizer2
+                        model_name = 'SASRec SMORL 2'
+
+                    main_QN.zero_grad()
+                    target_QN.zero_grad()
+
+                    target_Qs = target_QN.forward_next_states(
+                        next_state.to(device),
+                        len_next_state.to(device)
+                    )
+                    target_Qs_selector = main_QN.forward_next_states(
+                        next_state.to(device),
+                        len_next_state.to(device)
+                    )
+                    acc_reward = torch.where(is_buy > 0, reward_buy, reward_click)
+                    supervised_out, smorl_loss = main_QN(
+                        state.to(device).long(),
+                        len_state.to(device).long(),
+                        action.to(device),
+                        acc_reward.to(device),
+                        target_Qs.to(device),
+                        target_Qs_selector.to(device),
+                        is_done.to(device)
+                    )
+                    supervised_loss = criterion(supervised_out, action.to(device).long())
+                    loss = supervised_loss + smorl_loss
+                    scalar_loss = loss.item()
+
+                    optimizer.zero_grad()
+                    loss.backward()
+                    optimizer.step()
+
+                    total_step += 1
+
+                    if total_step % 200 == 0:
+                        print('Model is ', model_name)
+                        print('Supervised loss is %.3f, SMORL loss is %.3f' % (supervised_loss.item(), smorl_loss))
+                        print('Epoch {}......Step: {}....... Loss: {}'.format(epoch, total_step, scalar_loss = loss.item()))
+                        sys.stdout.flush()
+
+                    if total_step % args.eval_iter == 0:
+                        print('Evaluating Main Model')
+                        val_acc_main, predictions_main = evaluate(main_QN, args, 'val', state_size, item_num, device, args.div_emb_matrix, skip=args.skip, RL=args.rl)
+                        val_acc_main, predictions_main = evaluate(main_QN, args, 'test', state_size, item_num, device,
                                                                   args.div_emb_matrix, skip=args.skip, RL=args.rl)
-                    predictions_target.to_csv(
-                        '{}pred_target_acc{}_div{}_nov{}_epoch{}_step{}.tsv'.format(args.results_path, smorl_weights[0],
-                                                                                    smorl_weights[1], smorl_weights[2],
-                                                                                    epoch, total_step), sep='\t',
-                        index=False)
-                    print("I've saved pred tarhet at {}pred_target_acc{}_div{}_nov{}_epoch{}_step{}.tsv".format(
-                        args.results_path,
-                        smorl_weights[0], smorl_weights[1], smorl_weights[2], epoch, total_step))
-                    target_QN.train()
-                    print('Current accuracy of target model: ', val_acc_target)
-
-            current_time = time.time()
-            print('Epoch {}/{} Done'.format(epoch, args.epochs))
-            print('Total Time Elapsed: {} seconds'.format(str(current_time - start_time)))
-            epoch_times.append(current_time - start_time)
-        print('Total Training Time: {} seconds'.format(str(sum(epoch_times))))
-        sys.stdout.close()
-        sys.stdout = sys.__stdout__
+
+                        predictions_main.to_csv(f'{args.results_path}pred_main_{experiment_name}_epoch{epoch}_step{total_step}.tsv',
+                                                sep='\t', index=False)
+                        print(f"I've saved pred main at {args.results_path}pred_main_{experiment_name}_epoch{epoch}_step{total_step}.tsv")
+                        main_QN.train()
+                        print('Current accuracy of main model: ', val_acc_main)
+                        print('Evaluating Target Model')
+                        val_acc_target, predictions_target = evaluate(target_QN, args, 'val', state_size, item_num, device, args.div_emb_matrix, skip=args.skip, RL=args.rl)
+                        val_acc_target, predictions_target = evaluate(target_QN, args, 'test', state_size, item_num, device,
+                                                                      args.div_emb_matrix, skip=args.skip, RL=args.rl)
+                        predictions_target.to_csv(f'{args.results_path}pred_targetd_{experiment_name}_epoch{epoch}_step{total_step}.tsv',
+                                                  sep='\t', index=False)
+                        print(f"I've saved pred target at {args.results_path}pred_target_{experiment_name}_epoch{epoch}_step{total_step}.tsv")
+                        target_QN.train()
+                        print('Current accuracy of target model: ', val_acc_target)
+
+                current_time = time.time()
+                print('Epoch {}/{} Done'.format(epoch, args.epochs))
+                print('Total Time Elapsed: {} seconds'.format(str(current_time - start_time)))
+                epoch_times.append(current_time - start_time)
+            print('Total Training Time: {} seconds'.format(str(sum(epoch_times))))
+            sys.stdout.close()
+            sys.stdout = sys.__stdout__
